### ISSUE TITLE
Unable to reproduce results for OASIS dataset

### ISSUE THREAD

**Issue by iamsharmatul** [cwmok/LapIRN#11]
Dear @cwmok,

I am working on my thesis. In which I am evaluating multiple image registration networks. I have trained the `Train_LapIRN_diff.py` network for 60,000 iterations. But I am getting a dice score of 0.10 for validation as well as testing. Any inputs from your side will be really helpful. I will wait for your response. Thanks to you in advance.

Hyperparameters:
- lr = 1e-4
- antifold = 0
- smooth = 4
- freeze_step = 2000
- start_channel = 7

Training data:
- https://github.com/adalca/medical-datasets/blob/master/neurite-oasis.md
- I generate pairs and saved them in a `.csv `file and load this` .csv` fil in my training generator  
- All 3 levels are trained on 60,000 training samples. These samples are computed using combinations of (414) subjects without 
  repetition.
  
 Code:
 - I have attached the code below for your reference could you please help where am I going wrong
   [code.zip](https://github.com/cwmok/LapIRN/files/8215044/code.zip)
 
Training, Validation & Loss Plot:
 
Training loss for all levels:
![image](https://user-images.githubusercontent.com/52601678/157452265-473ced7d-5e7e-4156-a767-3de371d0b0b0.png)
![image](https://user-images.githubusercontent.com/52601678/157452412-9e517573-d3fe-41cf-aa77-4140e3d96129.png)
![image](https://user-images.githubusercontent.com/52601678/157452684-554ad3ae-87df-4dc4-8971-222d3649d127.png)
![image](https://user-images.githubusercontent.com/52601678/157452761-08db3212-78e6-481a-b047-79e909ffecf9.png)
![image](https://user-images.githubusercontent.com/52601678/157452891-c3e6ed22-e7d5-490f-87e7-5a6da8e4103d.png)

**Comment by cwmok** [comment]
Hi @iamsharmatul,

Your code seems good to me. I have modified a little bit and only run the training of level 3. The result is normal as shown below:

_Training lvl3...
epoch "0" step "0" -> training loss "-0.1843" - sim_NCC "-0.199932" - Jdet "1.0000098944" -smo "0.0039"Dice:  0.5978229121838816
epoch "0" step "1000" -> training loss "-0.2339" - sim_NCC "-0.258973" - Jdet "0.9996483922" -smo "
0.0063"Dice:  0.662033945014307_

I have attached the code with this message. [debug_code.zip](https://github.com/cwmok/LapIRN/files/8220905/debug_code.zip) Could you please try it out to see whether you can reproduce this result?

If you can reproduce the result, it implies there is something wrong with your "Dataset generator". If you can't, I think it may be due to a non-trivial environmental issue as discussed in https://github.com/cwmok/Fast-Symmetric-Diffeomorphic-Image-Registration-with-Convolutional-Neural-Networks/issues/10.

**Comment by iamsharmatul** [comment]
Dear @cwmok 

Thank you so much for giving your valuable time and look into this issue.

- I am able to reproduce the results from the shared debug code. 
![image](https://user-images.githubusercontent.com/52601678/157648946-5995773a-dcd2-4e7e-ae9c-b5b3f3ffae19.png)

- I will now carefully debug the data generators. And update the solution here.

**Comment by cwmok** [comment]
Hi @iamsharmatul,

I don't think your result is normal, as the dice score is not increasing in the first 3000 iterations. Could you provide your CUDA version and the version of Pytorch in your system? I would like to reproduce your results.

**Comment by iamsharmatul** [comment]
Dear @cwmok 

You are right the dice score is not increasing.
![image](https://user-images.githubusercontent.com/52601678/157681972-10a2f4ff-b502-443b-b01e-fed2fe6d694f.png)

I am currently running the code on ubuntu operating system, with Nvidia A100 GPU

MiniConda environment:
- pytorch 1.10.2  
- cuda = 11.3
- cudnn = 8.2.0_0
- python = 3.8.12

**Comment by iamsharmatul** [comment]
Dear @cwmok,

I have not tried VoxelMorph yet but I trained ConditionalLapirn with default hyperparameters on the OASIS dataset. The dice score obtained after training of 60,000 iterations is 0.13

**Comment by cwmok** [comment]
Perhaps you should try VoxelMorph first to make sure your dataset/learning environment is fine. At this moment, I cannot reproduce your result and I have no clue why this is happening. Feel free to let me know if you have any updates.

**Comment by cwmok** [comment]
I faced a similar issue today. I fixed it by switching to a different Pytorch CUDA version.
